### DISCUSSION TITLE
Scripting OperationMove to work on individual layers

### DISCUSSION THREAD

**Discussion by nmilnercollabora** [sn4k3/UVtools#509] [Scripts]
Hello,

I'm trying to write a script that copies and pastes a subset of layers which are not sequential. I was thinking I could use a for loop to run OperationMove by layer but I'm getting weird results. To simplify things I removed the for loop and tried the below code which should in my mind copy and move layers 20 and then 22. However, the first one works fine but the second results in a copy of a copy type view. What am I doing wrong?

Thanks

Nick

Layer 20 Original

![Screenshot from 2022-07-13 16-11-02](https://user-images.githubusercontent.com/10811764/178768400-6959adf7-eeca-4169-8cf9-6314206483d3.png)

Layer 20 after OperationMove

![Screenshot from 2022-07-13 16-10-42](https://user-images.githubusercontent.com/10811764/178769270-9d6ed9d0-9431-42e4-b9ec-0cf0ef7a6be3.png)

Layer 22 before

![Screenshot from 2022-07-13 16-11-02](https://user-images.githubusercontent.com/10811764/178768749-b9835f00-0b93-4f74-89d9-fc8ee2edc26b.png)

Layer 22 after OperationMove

![Screenshot from 2022-07-13 16-10-35](https://user-images.githubusercontent.com/10811764/178768823-c3d56c72-76fd-4ef2-b3d9-13c65081577d.png)

```
        new OperationMove(SlicerFile) 
	    {
	      IsCutMove = false,
	      MarginLeft = 400,
	      LayerIndexStart = 20,
	      LayerIndexEnd = 20,
	      
	    }.Execute(Progress);
	    
	    
	    new OperationMove(SlicerFile) 
	    {
	      IsCutMove = false,
	      MarginLeft = 400,
	      LayerIndexStart = 22,
	      LayerIndexEnd = 22,
	    }.Execute(Progress);

```

**Comment by nmilnercollabora** [comment]
I also tried using Tools -> Move where I did it copy and paste of all layers with margin left = 400 which did what was expected. I then repeated this exact same command and the result was the same copy overlapping copy view that I have seen above on each layer.

**Comment by sn4k3** [comment]
The Move tool require something to be selected to move. If you don't give the rectangle region you want to move/copy it will auto set to file `BoundingRectangle`, see: https://github.com/sn4k3/UVtools/blob/54d0cc1776282cf951861d71c7b35caed15870e2/UVtools.Core/Operations/OperationMove.cs#L187
https://github.com/sn4k3/UVtools/blob/54d0cc1776282cf951861d71c7b35caed15870e2/UVtools.Core/Operations/OperationMove.cs#L192
The first time you run the tool the `BoundingRectangle` it's equal to your single tile. But the second time you do `new OperationMove` the ROI will be the sum of two tiles, and then you are moving / copying a blank area that overlaps in the layer. 

To solve this you need to strict select the area you just want to copy OR if the copy area it's always the same, dont do a `new OperationMove` instead reuse one, eg:

```C#
var operation = new OperationMove(SlicerFile) 
	    {
	      IsCutMove = false,
	      MarginLeft = 400,
	    };

for(int layerIndex = startLayer; layerIndex  < endLayer; layerIndex ++)
{
   operation.SelectCurrentLayer(layerIndex);
   operation.Execute(Progress);
}
```

**Reply by nmilnercollabora** [reply]
Thank you - now I see the code that makes sense.

**Reply by sn4k3** [reply]
However if you want to copy the exact same number in all layers, use the Pattern tool instead to make copies over the plate

**Comment by nmilnercollabora** [comment]
@sn4k3 : The pattern option is greyed out in the Tools menu for me - not sure why. Is it possible to obtain the build plate width and height in pixels from within a script?

**Reply by sn4k3** [reply]
On UI i grey out the menus that are unable to run. But pattern tool may be able after generate testes. The reason why it is greyed is because first time file opens it finds there are not enough space to pattern and then it get disabled and wont check again, even if you free space... One fix to do it via UI is: Generate test, save, reload file, now pattern tool is enabled.

> Is it possible to obtain the build plate width and height in pixels from within a script?

`SlicerFile.Resolution` (display pixels) and `SlicerFile.Display` (display mm), consult this and below: 
https://github.com/sn4k3/UVtools/blob/54d0cc1776282cf951861d71c7b35caed15870e2/UVtools.Core/FileFormats/FileFormat.cs#L1475
https://github.com/sn4k3/UVtools/blob/54d0cc1776282cf951861d71c7b35caed15870e2/UVtools.Core/FileFormats/FileFormat.cs#L1507

**Comment by nmilnercollabora** [comment]
Thank you - is it possible to print debug strings from scripts (pop-up box or something displayed within the window) ?

**Reply by sn4k3** [reply]
You can show text on script window before execute.
All you can do to debug during or after is to write text to a file on your disk, using regular C# file acess and functions to write to text file.
eg:

```C#
using TextWriter writer = new StreamWriter(fullPath)); // Put on desktop for example - debug.txt
writer.WriteLine("Monica Rathbun");  
writer.WriteLine("Vidya Agarwal");  
writer.WriteLine("Mahesh Chand");  
writer.WriteLine("Vijay Anand");  
writer.WriteLine("Jignesh Trivedi");  
```

If you want to write text in a Parallel function, you need to use the async methods or lock one object first

You can also set `Operation.Tag` to display a message after script executation (i think)

**Comment by nmilnercollabora** [comment]
Ok thanks - I've been writing to the script window so far but yeah, writing to a file would also work.

**Comment by nmilnercollabora** [comment]
@sn4k3 : I've got the image (build plate) resolution is pixels now, how do I find the max pixel x/y for the actual part (green boundary box) ?

**Reply by sn4k3** [reply]
`SlicerFile.BoundingRectangle`
https://github.com/sn4k3/UVtools/blob/54d0cc1776282cf951861d71c7b35caed15870e2/UVtools.Core/FileFormats/FileFormat.cs#L1437

PS: Use the reply to anwser instead of open a new anwser here on discussions (If related to the same question/anwser)

**Reply by nmilnercollabora** [reply]
I do not have much experience programming and I can't figure out from the link below how to get the x y pixels of the boundary box from the SlicerFile.BoundingRectangle object. If you could give me another clue that would be very helpful.

**Reply by sn4k3** [reply]
`SlicerFile.BoundingRectangle` returns an C# rectangle structure, see and read here: https://docs.microsoft.com/en-us/dotnet/api/system.drawing.rectangle?view=net-6.0#properties
Tell me if by that you can get to solution

**Reply by nmilnercollabora** [reply]
Brilliant - all sorted now.
